### PR TITLE
feat(scripts): support OpenRC init system

### PR DESCRIPTION
**Pull request by mrl5**
### Description
introduce OpenRC bootup scripts

### How Has This Been Tested?
I've tested it by running
```console
# make uninstall
# lsmod | grep lkrg
```
```console
# make install &&
  /etc/init.d/lkrg start &&
  lsmod | grep lkrg &&
  /etc/init.d/lkrg status &&
  rc-update add lkrg boot &&
  rc-update show | grep lkrg
```
```console
# make uninstall
# lsmod | grep lkrg
# /etc/init.d/lkrg status
# rc-update show | grep lkrg
```
on OpenRC system ([funtoo linux](https://www.funtoo.org))

**update**
after gathering requirements in [this comment](https://github.com/lkrg-org/lkrg/pull/197#issuecomment-1175588110) I've additionally run
```console
# runlevel
N 3
# /etc/init.d/lkrg stop
# lsmod | grep lkrg;  echo $?
1
# /etc/init.d/lkrg start
# lsmod | grep lkr
p_lkrg                221184  0
```

```console
# init 1
# lsmod | grep lkr
p_lkrg                221184  0
```

and ensured that `p_lkrg` module is not unloaded from the kernel on poweroff or reboot

### CODE REVIEW

File: README [hunk 1]
```diff
@@ -116,39 +116,92 @@ so that it can then be loaded using the same procedure that's used on system
...
+
+	sudo mkdir -p /etc/modules-load.d/ &&
+		echo p_lkrg > /etc/modules-load.d/p_lkrg.conf
+
+Alternatively you can put the "modprobe p_lkrg" command into a system
+startup script.
+
+
+Uninstalling
+------------
```
Comment (solardiz) [inline README:185]: You'll need to rename and revise the "Uninstalling and upgrading" section then - it should probably become just "Upgrading" and should refer to the section "Uninstalling" instead of partially repeating its content.

File: README [hunk 2]
```diff
@@ -116,39 +116,92 @@ so that it can then be loaded using the same procedure that's used on system
...
+
 You can uninstall LKRG using "make" as well (still in the same directory):
 
 	sudo make uninstall
 
 You can also use the following command to temporarily stop the LKRG service
 without uninstalling it:
 
+for systemd:
```
Comment (solardiz) [inline README:194]: Style: you don't follow these with colons in your additions to other sections, so either should not do so here or should revise all of them. I'd do the latter, e.g.:

```
You can also use the following command to temporarily stop the LKRG service
without uninstalling it, for systemd:
```

and then it's natural to have the rest just like you wrote them here in `Uninstalling`.

File: README [hunk 3]
```diff
@@ -116,39 +116,92 @@ so that it can then be loaded using the same procedure that's used on system
...
+If your Linux distribution uses supported init system, you can install LKRG
+with:
 
 	sudo make install
 
-while you're still in its top level source code directory.
+while you're still in its top level source code directory. Meanwhile, on a
+system without supported init systems you can let "sudo make install" partially
+complete (up to the point where it finds you're not using supported init
+system).
```
Comment (solardiz) [inline README:144]: Should be "a supported init system" (add "a ") in both places. Also, two spaces before "Meanwhile" for consistent style with the rest of the file.

File: README [hunk 4]
```diff
@@ -116,39 +116,92 @@ so that it can then be loaded using the same procedure that's used on system
...
+for other
+
+	sudo modprobe -v p_lkrg
+
+
+Autoload on bootup
+------------------
+
+In order to automatically add LKRG module into Linux Kernel on each bootup run
+following command:
```
Comment (solardiz) [inline README:165]: Please replace this sentence with:

```
In order to automatically load LKRG into the Linux kernel on each bootup run
the following command:
```

File: scripts/bootup/openrc/lkrg [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#!/sbin/openrc-run
+# Distributed under the terms of the GNU General Public License v2
```
Comment (solardiz) [inline scripts/bootup/openrc/lkrg:2]: Why did you feel it important to specify license in this one tiny file and not in others? We do have top-level `LICENSE` specify GPLv2, so this feels unneeded (would have made sense if you specified a more relaxed license here).

File: README [hunk 6]
```diff
@@ -177,6 +230,10 @@ for LKRG with systemd, you may disable LKRG by specifying "nolkrg" on the
 kernel command-line via your bootloader.  The system should then boot up
 without LKRG, and thus without triggering the problem, letting you fix it.
 
+Alternative way (e.g. with OpenRC) is to use "module_blacklist=p_lkrg" kernel
+parameter. You must be aware though, that you will not be able to manually load
+the LKRG module if kernel was booted with this paramter.
```
[the remark this answers is not in the document]
Comment (mrl5) [reply README:235]: yup, I've checked and confirm that `p_lkrg` doesn't want to load with this option :)

**Comment by solardiz** [comment]
@mrl5 Also, you seem to have a GitHub-provided noreply e-mail address on the commit. You might want to change that to an address of your own, which you'd then add to your GitHub account.

**Comment by solardiz** [comment]
@mrl5 Is there any way for me to contact you about the misunderstandings I see in comments at https://bugs.funtoo.org/browse/FL-10115? That would be off-topic to discuss on this pull request.

**Comment by mrl5** [comment]
> @mrl5 Is there any way for me to contact you about the misunderstandings I see in comments at https://bugs.funtoo.org/browse/FL-10115? That would be off-topic to discuss on this pull request.

I've sent PM to your twitter account: https://twitter.com/solardiz

File: README (hunk 4 above)
Comment (mrl5) [reply README:165]: > Please replace this sentence with:
> 
> ```
> In order to automatically load LKRG into the Linux kernel on each bootup run
> the following command:
> ```

ok will do. for the record I've taken `add` from `man modprobe` :D
>       modprobe - Add and remove modules from the Linux Kernel

File: scripts/bootup/openrc/lkrg (hunk 5 above)
Comment (mrl5) [reply scripts/bootup/openrc/lkrg:2]: that's a good question - actually I did it per analogy to most of the scripts that are in `/etc/init.d/*` on my OS:
```console
$ head -n 3 /etc/init.d/* | grep -i license | wc -l
61
$ ls -l /etc/init.d | wc -l
106
```
but I can remove this line if you really want to

File: CHANGES [hunk 7]
```diff
@@ -1,3 +1,7 @@
+The following major changes have been made between LKRG 0.9.3 and 0.9.4:
```
Comment (solardiz) [inline CHANGES:1]: We generally only include the next version number in `CHANGES` shortly before making the release, and until then use something like "since LKRG 0.9.3".

File: README [hunk 8]
```diff
@@ -116,49 +116,100 @@ so that it can then be loaded using the same procedure that's used on system
...
 
 
+Supported init systems
+----------------------
+
+Currently there are dedicated scripts for:
+
+- systemd
+
+- OpenRC
```
Comment (solardiz) [inline README:126]: I'd shorten this to one line:

```
Currently there are dedicated scripts for systemd and OpenRC.
```

File: README [hunk 9]
```diff
@@ -116,49 +116,100 @@ so that it can then be loaded using the same procedure that's used on system
...
 
-If your Linux distribution uses systemd, you can install LKRG with:
+If your Linux distribution uses a supported init system, you can install LKRG
+with:
 
 	sudo make install
 
-while you're still in its top level source code directory.
+while you're still in its top level source code directory.  Meanwhile, on a
+OS without a supported init system you can let "sudo make install" partially
```
Comment (solardiz) [inline README:144]: Should be "an" before "OS", or use "distribution" in place of "OS". (I understand you don't want to write "system" referring to two different things in the same sentence.)

File: README [hunk 10]
```diff
@@ -116,49 +116,100 @@ so that it can then be loaded using the same procedure that's used on system
...
+complete (up to the point where it finds you're not using a supported init
+system).
+
+Run the following command to start the LKRG service:
 
-Run the following command to start the LKRG service just like it would be
-started on next system bootup:
+- for systemd:
```
Comment (solardiz) [inline README:150]: I'm fine with this style, but I'd save two lines by writing:

```
Run the following command to start the LKRG service, for systemd:
```

and then omit the leading "- " before "for OpenRC" and "for other".

Ditto in other places, so making the text significantly shorter.

File: README [hunk 11]
```diff
@@ -116,49 +116,100 @@ so that it can then be loaded using the same procedure that's used on system
...
-Our suggested way to upgrade LKRG is to start by uninstalling the old version,
-which you can do with:
-
-	sudo make uninstall
+Upgrading
+---------
 
+Our suggested way to upgrade LKRG is to start by uninstalling the old version,
 from the top level source code directory of that old version (doing so using
 the source code tree of the new version may or may not work correctly).
```
Comment (solardiz) [inline README:214]: I think this paragraph about the directory should stay in "Uninstalling", not be in "Upgrading", which should refer to "Uninstalling" for it.

File: README [hunk 12]
```diff
@@ -171,11 +222,12 @@ Recovery
...
-unit file for LKRG.  Thus, if you've followed the above installation procedure
-for LKRG with systemd, you may disable LKRG by specifying "nolkrg" on the
-kernel command-line via your bootloader.  The system should then boot up
-without LKRG, and thus without triggering the problem, letting you fix it.
+bootup, we've included support for the "nolkrg" kernel parameter.  Thus, you
+may disable LKRG by specifying "nolkrg" on the kernel command-line via your
+bootloader.  The system should then boot up without LKRG, and thus without
+triggering the problem, letting you fix it.  You must be aware though, that you
+will not be able to manually load the LKRG module if kernel was booted with
+this paramter.
```
Comment (solardiz) [inline README:230]: Add "the" before "kernel", fix typo in "paramter".

File: scripts/bootup/openrc/lkrg (hunk 5 above)
Comment (solardiz) [reply scripts/bootup/openrc/lkrg:2]: OK, up to you.

**Review by solardiz — CHANGES_REQUESTED** [review]
Looks almost ready.

**Comment by solardiz** [comment]
@mrl5 BTW, for the sysctl file to provide (almost-)initial LKRG configuration, LKRG needs to be loaded _before_ that file is processed. However, I don't see anything in the init script that would indicate its priority vs. other scripts. Did you look into that?

It's also desirable for security that LKRG be loaded before services such as sshd, cron, MTA/MDA, and unloaded late on system shutdown (if at all).

On pre-systemd Red Hat'ish systems with `chkconfig`, this would be controlled with lines like `# chkconfig: 2345 55 25` in init files (comments for the shell, but magic for `chkconfig`). These indicated runlevels (`2345` here) and start and stop priorities vs. other init scripts (`55` and `25` here). Does OpenRC have anything similar? Does Funtoo use that?

Quite possibly we'll eventually drop unloading of LKRG on system shutdown somehow. This is currently mostly a side-effect of our uninstall/upgrade instructions.

**Comment by mrl5** [comment]
> @mrl5 BTW, for the sysctl file to provide (almost-)initial LKRG configuration, LKRG needs to be loaded _before_ that file is processed (edit: otherwise those settings would not take effect at all). However, I don't see anything in the init script that would indicate its priority vs. other scripts. Did you look into that?
> 
> It's also desirable for security that LKRG be loaded before services such as sshd, cron, MTA/MDA, and unloaded late on system shutdown (if at all).
> 
> On pre-systemd Red Hat'ish systems with `chkconfig`, this would be controlled with lines like `# chkconfig: 2345 55 25` in init files (comments for the shell, but magic for `chkconfig`). These indicated runlevels (`2345` here) and start and stop priorities vs. other init scripts (`55` and `25` here). Does OpenRC have anything similar? Does Funtoo use that?
> 
> Quite possibly we'll eventually drop unloading of LKRG on system shutdown somehow. This is currently mostly a side-effect of our uninstall/upgrade instructions.

thanks for bringing this up. it actually made me to take a step back and ask this question: https://github.com/lkrg-org/lkrg/issues/109#issuecomment-1175514381

TBH the only reason why I decided to do this pull request was to make OpenRC service per analogy to the existing systemd one. But now I'm really thinking to give up this idea and simply fall back to [modules-load.d](https://www.freedesktop.org/software/systemd/man/modules-load.d.html).

I'm pretty sure OpenRC takes care of correct loading order when [modules-load.d](https://www.freedesktop.org/software/systemd/man/modules-load.d.html) is detected (like mentioned first sysctl file then modules as well as modules before other services). Additionally I don't think that it would unload the modules before shutdown.

If you agree with above statement I'd be happy to modify `README` to include some instructions for [modules-load.d](https://www.freedesktop.org/software/systemd/man/modules-load.d.html)

**Comment by solardiz** [comment]
> I'm pretty sure OpenRC takes care of correct loading order when [modules-load.d](https://www.freedesktop.org/software/systemd/man/modules-load.d.html) is detected

What makes you so sure?

> like mentioned first sysctl file then modules

For LKRG, it should be the other way around! The sysctl's can't be applied before the module is loaded, they simply don't exist yet. For some other module, it could be different - maybe that other module does not have its own sysctl's, but its loading is affected by the kernel's sysctl settings. So there isn't a universal "correct" order that an init system could use.

Anyway, longer term we should probably move away from usage of sysctl's for initial configuration. We intended module parameters to be used for that, but then got contributed this file with all sysctl's based on our `README`, and I understand it's more convenient to edit than a list of module parameters all on one line, but it's also less reliable and is inherently racy.

> I'd be happy to modify README to include some instructions for [modules-load.d](https://www.freedesktop.org/software/systemd/man/modules-load.d.html)

Then we also need instructions for passing parameters to modules loaded that way.

Since you started this PR with an almost ready contribution, I thought we could accept it for now, then _possibly_ drop both systemd and OpenRC specifics at some later time.

**Comment by mrl5** [comment]
> What makes you so sure?
It's an assumption :D sorry - it might sound misleading

> Since you started this PR with an almost ready contribution, I thought we could accept it for now, then possibly drop both systemd and OpenRC specifics at some later time.
very well then - let me go through 2nd iteration of your CR so that I can apply fixes. I'll click on `re-request review` once I think I'm ready

**Comment by mrl5** [comment]
@solardiz for my convenience I decided to gather requirements for proper way of handling `p_lkrg` module by the init system dedicated service (I'm using [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119) nomenclature):
1. it MUST load `p_lkrg` module before sysctl config files are processed (esp. `/etc/sysctl.d/01-lkrg.conf`)
2. it MUST load `p_lkrg` module as early after bootup as possible (esp. before services such as sshd, cron, MTA/MDA are started)
3. it MUST unload `p_lkrg` module on explicit user command
4. it SHOULD NOT unload `p_lkrg` module automatically at shutdown
5. it MAY not load `p_lkrg` module if `nolkrg` kernel parameter was provided

pls suggest changes to above if I missed sth or misunderstood sth.

**Comment by solardiz** [comment]
@mrl5 You got those requirements (1, 2, 3) and preferences (4, 5) right. The changes you suggest in "update" are also within consideration. Thanks!

File: scripts/bootup/openrc/lkrg [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+	ewarn "This command doesn't unload \"p_lkrg\" module from the kernel!"
+	# it SHOULD unload `p_lkrg` module on explicit user command
+	einfo "To unload \"p_lkrg\" module use \"unload\" target"
+}
+
+unload() {
+	ebegin "Unloading \"p_lkrg\" module from the kernel"
+	$command $unload_command_args
+	eend $?
+}
```
Comment (solardiz) [inline scripts/bootup/openrc/lkrg:31]: I'm OK with this approach, although I thought you could probably check `runlevel` from within `stop`. Is there a working `runlevel` command on systems with OpenRC?

File: scripts/bootup/openrc/lkrg-openrc.sh [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+	fi
+	if [ -e "$P_SYSCTL_DIR/01-lkrg.conf" ]; then
+		echo -e "       ${P_YL}01-lkrg.conf${P_GREEN} is already installed, skipping${P_NC}"
+	else
+		echo -e "       ${P_GREEN}Installing ${P_YL}01-lkrg.conf${P_GREEN} file under ${P_YL}$P_SYSCTL_DIR${P_GREEN} directory${P_NC}"
+		install -pm 644 -o root -g root scripts/bootup/lkrg.conf "$P_SYSCTL_DIR/01-lkrg.conf"
+	fi
+elif [ "$1" == "uninstall" ]; then
+	echo -e "       ${P_GREEN}Stopping ${P_YL}lkrg${P_NC}"
+	/etc/init.d/lkrg stop
```
[the remark this answers is not in the document]
Comment (mrl5) [reply scripts/bootup/openrc/lkrg-openrc.sh:41]: I've managed to move everything to `stop` command keeping [the requirements](https://github.com/lkrg-org/lkrg/pull/197#issuecomment-1175588110)

thanks for the `runlevel` hint in this comment https://github.com/lkrg-org/lkrg/pull/197#discussion_r916613571

File: scripts/bootup/openrc/lkrg [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+start() {
+	ebegin "Loading \"p_lkrg\" module to the kernel"
+	$command $command_args
+	eend $?
+}
+
+stop() {
+	# for more info about runlevels (e.g. "S", "0", "1", "6") check: man 8 init
+	if [[ $($runlevel_command) =~ (S|0|1|6)$ ]]; then
```
Comment (solardiz) [inline scripts/bootup/openrc/lkrg:22]: Maybe add a space before each runlevel value we check, so that we'd only match on the second token in `runlevel` output? At least for me, `runlevel` outputs two tokens - previous and current runlevels. Here, the previous can potentially be single-user mode, from which the sysadmin might have exited into multi-user. We probably do want to load LKRG in that case.

I mean:

```bash
	if [[ $($runlevel_command) =~ ( S| 0| 1| 6)$ ]]; then
```

Of course, you need to re-test it.

BTW, is it safe to assume the shell is `bash` here?

**Review by solardiz — APPROVED** [review]
This looks ready to me. I just raised one minor concern - please feel free to make an adjustment (and re-test) or let me know that you're leaving things as they are (and I'll merge as-is). Thanks!

**Comment by solardiz** [comment]
BTW, note that this PR looks like it'll clash with #200. I hope you don't mind rebasing if we happen to merge that other one first.

**Comment by mrl5** [comment]
> BTW, note that this PR looks like it'll clash with #200. I hope you don't mind rebasing if we happen to merge that other one first.

I don't mind rebasing in case of git conflict :) I hope you don't mind reviewing one more time in case this happens :D

File: scripts/bootup/openrc/lkrg (hunk 15 above)
Comment (mrl5) [reply scripts/bootup/openrc/lkrg:22]: > Maybe add a space before each runlevel value we check, so that we'd only match on the second token in `runlevel` output? At least for me, `runlevel` outputs two tokens - previous and current runlevels. Here, the previous can potentially be single-user mode, from which the sysadmin might have exited into multi-user. We probably do want to load LKRG in that case.

why not - I'll do it. For the record though, my understanding of `$` regex anchor, `man 8 init` and `man 8 runlevel` is that there shouldn't be any bug with current regex

> BTW, is it safe to assume the shell is bash here?

good point - better safe than sorry - actually I'm not sure how `#!/sbin/openrc-run` works under the hood

**Comment by solardiz** [comment]
Looks good to me. I'll merge now. Thank you!
